### PR TITLE
Almost remove EditorShell.Current

### PR DESCRIPTION
**Pull request by AlexanderSher**
The reason for this change is to avoid bugs like we had later in previous iteration when `EditorShell.Current` was null when it was first accessed and cached. Another reason is to allow editor tests to run concurrently.
The only two places where `EditorShell.Current` is still used are REditorSettings and RToolsSettings. Removing it from there require additional changes and can be done in a separate pull request.

### CODE REVIEW

**Comment by AlexanderSher** [comment]
Let's use Codeflow for this change - GitHub dies on big changes

File: src/UnitTests/Core/Impl/Threading/UIThreadHelper.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@
 using System.Windows.Threading;
 using Microsoft.Common.Core;
 using Microsoft.Common.Core.Tasks;
+using NSubstitute.Exceptions;
```
Comment (AlexanderSher) [inline src/UnitTests/Core/Impl/Threading/UIThreadHelper.cs]: Will remove this

**Comment by MikhailArkhipov** [comment]
>             EditorTree tree = new EditorTree(textBuffer);

Sounds line a merge conflict, I added using with the last merge

---
Refers to: src/R/Editor/Test/Signatures/ComputeCurrentParameterTest.cs:43 in 9069c6d. [](commit_id = 9069c6d7486a134450bf3b677b6c4b65a496e27b, deletion_comment = False)

**Comment by MikhailArkhipov** [comment]
>             using (var tree = new EditorTree(textBuffer)) {

As above

---
Refers to: src/R/Editor/Test/Signatures/ComputeCurrentParameterTest.cs:99 in 9069c6d. [](commit_id = 9069c6d7486a134450bf3b677b6c4b65a496e27b, deletion_comment = True)

File: src/R/Editor/Impl/Validation/Tagger/EditorErrorTagger.cs [hunk 2]
```diff
@@ -28,6 +29,8 @@ namespace Microsoft.R.Editor.Validation.Tagger {
     /// As the stylesheet changes, this class keeps the list of syntax errors up to date.
     /// </summary>
     public class EditorErrorTagger : ITagger<IErrorTag>, IEditorTaskListItemSource {
+        private readonly ICoreShell _shell;
+
         [Import(AllowDefault = true)]
         internal IEditorTaskList TaskList { get; set; }
```
Comment (MikhailArkhipov) [inline src/R/Editor/Impl/Validation/Tagger/EditorErrorTagger.cs]: Probably can be imported and passed down from the provider eliminating SatisfyImportsOnce

**Comment by MikhailArkhipov** [comment]
>         ITextBuffer _textBuffer;

private/readonly if you don't mind

---
Refers to: src/R/Editor/Impl/Validation/Tagger/EditorErrorTagger.cs:47 in 9069c6d. [](commit_id = 9069c6d7486a134450bf3b677b6c4b65a496e27b, deletion_comment = False)

File: src/R/Editor/Impl/Signatures/SignatureHelpSource.cs [hunk 3]
```diff
@@ -20,9 +21,11 @@
 namespace Microsoft.R.Editor.Signatures {
     sealed class SignatureHelpSource : ISignatureHelpSource {
         private ITextBuffer _textBuffer;
```
Comment (MikhailArkhipov) [inline src/R/Editor/Impl/Signatures/SignatureHelpSource.cs]: readonly

**Comment by MikhailArkhipov** [comment]
>             rcc?.TriggerSignatureHelp();

Merge conflict, should be the other way around

---
Refers to: src/R/Editor/Impl/Signatures/SignatureHelp2.cs:21 in 9069c6d. [](commit_id = 9069c6d7486a134450bf3b677b6c4b65a496e27b, deletion_comment = True)

File: src/R/Editor/Impl/Outline/ROutliningTagger.cs [hunk 4]
```diff
@@ -4,14 +4,15 @@
...
+using Microsoft.Languages.Editor.Shell;
 using Microsoft.R.Editor.Document;
 using Microsoft.VisualStudio.Text.Tagging;
 
 namespace Microsoft.R.Editor.Outline {
     internal sealed class ROutliningTagger : OutliningTagger {
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2000:Dispose objects before losing scope")]
         public ROutliningTagger(REditorDocument document)
-            : base(document.EditorTree.TextBuffer, new ROutlineRegionBuilder(document)) {
+            : base(document.EditorTree.TextBuffer, new ROutlineRegionBuilder(document, EditorShell.Current)) {
```
Comment (MikhailArkhipov) [inline src/R/Editor/Impl/Outline/ROutliningTagger.cs]: Can be imported in OutliningTaggerProvider

**Comment by MikhailArkhipov** [comment]
>             EditorTree.Closing += OnEditorTreeClosing;

Merge conflict

---
Refers to: src/R/Editor/Impl/Outline/ROutlineRegionBuilder.cs:52 in 9069c6d. [](commit_id = 9069c6d7486a134450bf3b677b6c4b65a496e27b, deletion_comment = True)

File: src/R/Editor/Impl/Signatures/SignatureHelpSource.cs (hunk 3 above)
Comment (AlexanderSher) [inline src/R/Editor/Impl/Signatures/SignatureHelpSource.cs]: Can't be readonly, it becomes `null` in `Dispose`. Do you want me to remove that part from `Dispose` method?

---
In reply to: [68842049](https://github.com/Microsoft/RTVS/pull/2022#discussion_r68842049) [](ancestors = 68842049)

Comment (MikhailArkhipov) [inline src/R/Editor/Impl/Signatures/SignatureHelpSource.cs]: Yeah, I don't think it really needs to be nullified.

Comment (AlexanderSher) [inline src/R/Editor/Impl/Signatures/SignatureHelpSource.cs]: Done

---
In reply to: [68853191](https://github.com/Microsoft/RTVS/pull/2022#discussion_r68853191) [](ancestors = 68853191)

File: src/R/Editor/Impl/Validation/Tagger/EditorErrorTagger.cs (hunk 2 above)
Comment (AlexanderSher) [inline src/R/Editor/Impl/Validation/Tagger/EditorErrorTagger.cs]: Fixed

---
In reply to: [68841881](https://github.com/Microsoft/RTVS/pull/2022#discussion_r68841881) [](ancestors = 68841881)

**Comment by AlexanderSher** [comment]
>         ITextBuffer _textBuffer;

Both `_textBuffer` and `_document` are set to null in `OnDocumentClosing`. Do we need it?

---
In reply to: [229181322](https://github.com/Microsoft/RTVS/pull/2022#issuecomment-229181322) [](ancestors = 229181322)

---
Refers to: src/R/Editor/Impl/Validation/Tagger/EditorErrorTagger.cs:47 in 9069c6d. [](commit_id = 9069c6d7486a134450bf3b677b6c4b65a496e27b, deletion_comment = False)

File: src/R/Editor/Impl/Outline/ROutliningTagger.cs (hunk 4 above)
Comment (AlexanderSher) [inline src/R/Editor/Impl/Outline/ROutliningTagger.cs]: What can be imported? `ROutlineRegionBuilder`?

---
In reply to: [68843223](https://github.com/Microsoft/RTVS/pull/2022#discussion_r68843223) [](ancestors = 68843223)

**Comment by AlexanderSher** [comment]
>             EditorTree.Closing += OnEditorTreeClosing;

Sorry, don't get it. There is no change on this line

---
In reply to: [229183336](https://github.com/Microsoft/RTVS/pull/2022#issuecomment-229183336) [](ancestors = 229183336)

---
Refers to: src/R/Editor/Impl/Outline/ROutlineRegionBuilder.cs:52 in 9069c6d. [](commit_id = 9069c6d7486a134450bf3b677b6c4b65a496e27b, deletion_comment = True)

**Comment by AlexanderSher** [comment]
>             EditorTree tree = new EditorTree(textBuffer);

Fixed in all the places where `EditorTree` is created

---
In reply to: [229176727](https://github.com/Microsoft/RTVS/pull/2022#issuecomment-229176727) [](ancestors = 229176727)

---
Refers to: src/R/Editor/Test/Signatures/ComputeCurrentParameterTest.cs:43 in 9069c6d. [](commit_id = 9069c6d7486a134450bf3b677b6c4b65a496e27b, deletion_comment = False)

**Comment by MikhailArkhipov** [comment]
👍

**Comment by int19h** [comment]
:shipit:

**Pull request merged**